DR2-1635 TDR package builder

MancunianSam opened the pull request:
There are some other changes here that I had trouble separating.

The AWS clients now seem to be doing multipart uploads. This is
something that's hidden in the transfer manager so we can't control it.

I've had to add/modify a lot of the wiremock stubs to get it to work
with the latest AWS libraries. This isn't great as they might change
their minds again later so I think Wiremock has to go but not yet.

Also, now they're on multipart uploads, the checksum we get back for the
metadata file in the parsed court document handler is not reliable so
I've added in a line to calculate checksums as they come out of the
tar.gz. It's slightly inefficient but I think we can live with it.

techncl reviewed (commented):
Looks good! Just some minor comments. Also, could you add a README for this?

techncl commented on scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:
```diff
@@ -40,7 +40,7 @@ class Lambda extends LambdaRunner[Input, Output, Config, Dependencies]:
         val fileId = dependencies.uuidGenerator()
         val metadataId = dependencies.uuidGenerator()
-        val archiveFolder = ArchiveFolderMetadataObject(archiveFolderId, None, None, tdrMetadata.ConsignmentReference, tdrMetadata.Series, Nil)
+        val archiveFolder = ContentFolderMetadataObject(archiveFolderId, None, None, tdrMetadata.ConsignmentReference, Nil)
```
The name is still `archiveFolder` and the id is still `archiveFolderId`

MancunianSam replied:
Ah yeah cheers, I've changed it.

MancunianSam commented on scala/lambdas/build.sbt:
```diff
@@ -20,6 +20,7 @@ lazy val ingestLambdasRoot = (project in file("."))
     ingestParsedCourtDocumentEventHandler,
     ingestUpsertArchiveFolders,
     ingestWorkflowMonitor,
+    preIngestTdrPackageBuilder,
```
Yep, I've put them in alphabetical order.

MancunianSam commented on scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:
```diff
@@ -0,0 +1,179 @@
+
+  override def handler: (Input, Config, Dependencies) => IO[Unit] = (input, config, dependencies) => {
+
+    def processTdrMetadata(tdrMetadataStream: Stream[IO, TDRMetadata], fileLocation: URI): Stream[IO, MetadataObject] = {
```
You're duplicating information that way. The argument is a `TDRMetadata` stream and the return type is a `MetadataObject` stream. You don't need to name the method like that as well. 
It's also quite difficult to parse the code when the names are too long so I'll leave it.

MancunianSam commented on scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:
```diff
@@ -0,0 +1,179 @@
+          tdrMetadata.TransferringBody,
+          LocalDateTime.parse(tdrMetadata.TransferInitiatedDatetime.replace(" ", "T")).atOffset(ZoneOffset.UTC),
+          "TDR",
+          "Born Digital",
```
As far as I'm aware there are no plans to bring in anything other than born digital through TDR so this will be fine for a good long while.

MancunianSam commented on scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:
```diff
@@ -0,0 +1,179 @@
+end Lambda
+
+object Lambda:
+  case class TDRMetadata(
```
Because they come in from TDR that way and I didn't want to have a custom decoder just to change the case of the variables.

MancunianSam commented on scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:
```diff
@@ -0,0 +1,179 @@
+      FileReference: String
+  )
+
+  case class LockTableMessage(id: UUID, location: URI)
```
If we change the lock table message then we'll have to change the case class name. And you can see it has an id and a location by the arguments so I think it's ok as is.

MancunianSam commented on scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:
```diff
@@ -0,0 +1,181 @@
+        .compile
+        .toList
+        .flatMap { metadata =>
+          IO.raiseWhen(metadata.isEmpty)(new Exception(s"Metadata list for ${input.groupId} is empty")) >> {
```
I don't think this is the place for that kind of validation. If it was to go anywhere it would go in your validation lambda but we'll see if we get any issues like that first.

techncl replied:
Ok. Yes, it is part of the validation lambda

MancunianSam commented on scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:
```diff
@@ -0,0 +1,181 @@
+        val assetId = tdrMetadata.UUID
+        val fileId = dependencies.uuidGenerator()
+        val metadataId = dependencies.uuidGenerator()
+        val archiveFolder = ArchiveFolderMetadataObject(archiveFolderId, None, None, tdrMetadata.ConsignmentReference, tdrMetadata.Series, Nil)
```
You've commented on this below but this shouldn't have been an archive folder, that's a weird judgment only thing so I've changed it to a content folder. We might not keep it anwyay but it's fine for now.

MancunianSam commented on scala/lambdas/preingest-tdr-package-builder/src/main/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/Lambda.scala:
```diff
@@ -0,0 +1,181 @@
+      tdrMetadataStream.flatMap { tdrMetadata =>
+        val archiveFolderId = dependencies.uuidGenerator()
+        val assetId = tdrMetadata.UUID
+        val fileId = dependencies.uuidGenerator()
```
We won't, one metadata file from TDR is one file from TDR and that's not likely to change.

MancunianSam commented:
I've added in a README. I've also added a series attribute to the `ContentFolderMetadataObject` because the ingest mapper needs it to find the correct series.

techncl reviewed (approved):
👍
